File: BOJ/2583.kt
import java.io.*
import java.util.*
import kotlin.Pair as Pair

fun main() {
    val br = BufferedReader(InputStreamReader(System.`in`))
    val (m, n, k) = br.readLine().split(' ').map{it.toInt()}
    var matrix = Array(m, {IntArray(n,{0})})
    var checked = Array(m, {IntArray(n,{0})})
    var row_dir = arrayOf(1,0,-1,0)
    var col_dir = arrayOf(0,-1,0,1)
    var cnt = 0
    var areas = mutableListOf<Int>()
    var queue = arrayListOf<Pair<Int,Int>>()

    for (a in 0..k-1){
        val (x1, y1, x2, y2) = br.readLine().split(' ').map{it.toInt()}
        for (i in y1..(y2-1)){
            for (j in x1..(x2-1)){
                matrix[i][j] = 1
            }
        }
    }

    for (i in 0..m-1){
        for (j in 0..n-1){
            if(matrix[i][j] == 0 && checked[i][j] == 0){
                cnt += 1
                var area = 1
                checked[i][j] = 1
                queue.add(Pair(i,j))

                while(queue.isNotEmpty()){
                    val (cur_row, cur_col) = queue.removeAt(0)
                    for (k in 0..3){
                        val next_row = cur_row + row_dir[k]
                        val next_col = cur_col + col_dir[k]
                        if ((next_row>=0 && next_row<m) && (next_col>=0 && next_col<n)){
                            if (matrix[next_row][next_col] == 0 && checked[next_row][next_col]==0){
                                checked[next_row][next_col] = 1
                                queue.add(Pair(next_row, next_col))
                                area += 1
                            }
                        }
                    }
                }
                areas.add(area)
            }
        }
    }
    areas.sort()
    println(cnt)
    for(a in areas){
        print("$a ")
    }
}

File: BOJ/2583.py
from collections import deque
import sys

input = sys.stdin.readline

m, n, k = map(int,input().split())
answer = []
graph = [[0]*(n) for _ in range(m)]
checked = [[False]*(n) for _ in range(m)]
region_cnt = 0

x = [-1,1,0,0]
y = [0,0,-1,1]

for _ in range(k):
    x1,y1,x2,y2 = map(int,input().split())                   
    x3, y3 = x1, y2
    x4, y4 = x2, y1
    for col in range(m-y3 ,m-y4):
        for row in range(x3, x4):
            graph[col][row] = 1

for start_col in range(m):
    for start_row in range(n):
        square = 0
        if graph[start_col][start_row] == 0 and checked[start_col][start_row] == False:
            region_cnt += 1
            square += 1
            q = deque()
            q.append((start_col, start_row))
            checked[start_col][start_row] = True

            while q:
                cur_col, cur_row = q.popleft()
                for i in range(4):
                    next_col = cur_col+x[i]
                    next_row = cur_row+y[i]
                    if next_col >= 0 and next_col < m and next_row >= 0 and next_row < n:
                        if graph[next_col][next_row]==0 and checked[next_col][next_row] == False:
                            square += 1
                            q.append((next_col, next_row))
                            checked[next_col][next_row] = True

            else:
                answer.append(square)

answer.sort()
             
print(region_cnt)
print(*answer)